cabal-install/src/Distribution/Client/CmdRepl.hs

```diff
@@ -285,15 +287,56 @@ multiReplDecision ctx compiler flags =
 -- For more details on how this works, see the module
 -- "Distribution.Client.ProjectOrchestration"
 replAction :: NixStyleFlags ReplFlags -> [String] -> GlobalFlags -> IO ()
-replAction flags@NixStyleFlags{extraFlags = r@ReplFlags{..}, ..} targetStrings globalFlags =
-  withContextAndSelectors verbosity AcceptNoTargets (Just LibKind) flags targetStrings globalFlags ReplCommand $ \targetCtx ctx targetSelectors -> do
+replAction flags@NixStyleFlags{extraFlags = r@ReplFlags{..}, ..} targetStrings' globalFlags = do
+  -- NOTE: The REPL will work with no targets in the context of a project if a
+  -- single package is in the same directory as the project file. To have the
+  -- same behaviour when the package is somewhere else we adjust the targets.
+  targetStrings <-
+    if null targetStrings'
+      then withCtx silent targetStrings' $ \targetCtx ctx _ ->
+        return . fromMaybe [] $ case targetCtx of
+          ProjectContext ->
+            let pkgs = projectPackages $ projectConfig ctx
+             in if length pkgs == 1
+                  then pure <$> listToMaybe pkgs
+                  else Nothing
+          _ -> Nothing
+      else return targetStrings'
+
+  withCtx verbosity targetStrings $ \targetCtx ctx targetSelectors -> do
     when (buildSettingOnlyDeps (buildSettings ctx)) $
       dieWithException verbosity ReplCommandDoesn'tSupport
     let projectRoot = distProjectRootDirectory $ distDirLayout ctx
         distDir = distDirectory $ distDirLayout ctx
 
     baseCtx <- case targetCtx of
-      ProjectContext -> return ctx
+      ProjectContext -> do
+        let pkgs = projectPackages $ projectConfig ctx
+        when (null targetStrings && length pkgs /= 1) $
```

Comment:
At this point, should you not be looking at `targetSelectors`, the resolved target strings?